# scDamAndTools

**scDamAndTools** is a set of functions for the processing of raw data generated by scDam&T-seq, a method to simultaneously measure protein-DNA interactions and transcription from single cells (Rooijers _et al_., 2019). scDam&T-seq combines a DamID-based method to measure protein-DNA interactions and an adaptation of CEL-Seq to measure transcription. The starting point of the workflow is raw sequencing data and the end result are tables of UMI-unique DamID and CEL-Seq counts.

## Installation
**scDamAndTools** was developed in Python 3.6.5 and is compatible with Python 3.6+.

### Setting up a virtual environment
We recommend users to install scDamAndTools and its requirements in a Python virtual environment. In order to generate a virtual environment:

```
python3 -m venv $HOME/.venvs/my_venv
```

To activate the virtual environment:
```
source ~/.venvs/my_venv/bin/activate
```

### Requirements
Install necessary packages:

```
pip install --upgrade pip wheel setuptools
pip install cython
```

### Installing scDamAndTools
```
pip install git+https://github.com/KindLab/scDamAndTools.git
```

## Usage
**scDamAndTools** contains eight functions for the processing of raw sequencing data generated by scDam&T-seq experiments:  
|function|description|
|---|---|
|`add_read_prefix.awk`|Appends additional bases (e.g. "GA") to the beginning of reads.|
|`bin_damid_counts.py`|Bins UMI-unique DamID counts into genomically equal-sized bins.|
|`demultiplex.py`|Demultiplexes all reads in a library based on their barcode.|
|`fetch_regions.py`|Generates _in silico_ reads of a defined read length for all motif occurrences.|
|`find_motif_occurrences.py`|Finds motif (e.g. "GATC") occurrences genome-wide from a FASTA file.|
|`generate_celseq_counts.py`|Processes aligned CEL-Seq reads into a table of UMI-unique counts.|
|`generate_damid_counts.py`|Processes aligned DamID reads into a table of UMI-unique counts.|
|`write_posarray.py`|Generates a table with the positions of a motif genome-wide.|

### add_read_prefix.awk
```
add_read_prefix.awk -v PREFIX="GA" infile
```

Takes a FASTQ file as an input and appends the specified prefix to the beginning of all reads. The output in FASTQ format and is directed to STDOUT.

|arguments|option|description|
|---|---|---|
|`infile`|required|FASTQ file with sequence reads. Alternatively, reads can be piped via STDIN.|
|`-v PREFIX=`|required|Sequence to be appended to the beginning of all reads.|

### bin_damid_counts.py
```
bin_damid_counts.py [-h] [--verbose] [--quiet] --outfile OUTFILE [--binsize INT] --posfile HDF5_FILE [--mapfile HDF5_FILE] FILE [FILE ...]
```

Takes one or more position-wise UMI-unique DamID count tables (_HDF5_) as an input and outputs a binned count table with binsizes of the specified size. All input files are summed in the final output file.

|arguments|option|description|
|---|---|---|
|`infiles`|required|HDF5 file(s) of position-wise UMI-unique DamID count table(s).|
|`--outfile`|required|HDF5 file name to write output to.|
|`--binsize`|optional|Size of the genomic bins into which data will be summarized [default: 2000].|
|`--posfile`|required|Location of the motif position array (HDF5).|
|`--mapfile`|required|Location of the motif mappability array (HDF5).|
|`-v/--verbose`|optional|Set level of logging messages.|
|`-q/--quiet`|optional|Silence all logging messages.|

### demultiplex.py
```
demultiplex.py [-h] [--verbose] [--quiet] -o OUTFMT [-m MISMATCHES] [-k] [--verify-no-ambiguous] [--unmatched-outfile FILE] [--ambiguous-outfile FILE] [--infofile FILE] bcfile infiles [infiles ...]
```
Takes one or more raw sequencing files (_FASTQ_) as an input and outputs a file (_FASTQ_) for each DamID/CEL-Seq barcode provided in the barcode file containing all reads matching that barcode.

|arguments|option|description|
|---|---|---|
|`bcfile`|required|TAB-separated file with DamID and CEL-Seq barcode names and barcode sequences corresponding to the samples in the library.|
|`infiles`|required|FASTQ files of raw sequencing reads.|
|`-o/--outfmt`|required|File name format for output files with demultiplexed reads per sample. Should contain a "{name}" field for the barcode name (and a {readname} field for PE input).|
|`-m/--mismatches`|optional|The maximum number of mismatches allowed between the provided barcodes and observed sequences [default: 1].|
|`-k/--keep`|optional|If set, the barcode sequences are not trimmed from the reads [default: False].|
|`--verify-no-ambiguous`|optional|If set, verifies whether two or more provided barcode sequences are inherently ambiguous given the number of allowed mismatches [default: False].|
|`--unmatched-outfile`|optional|Output file (BAM) for reads in the library that could not be matched to any barcode. In the case of PE input, should contain a "{readname}" field.|
|`--ambiguous-outfile`|optional|Output file (BAM) for reads in the library that match multiple barcodes. In the case of PE input, should contain a "{readname}" field.|
|`--infofile`|optional|File in which to report demultiplexing stats.|
|`-v/--verbose`|optional|Set level of logging messages.|
|`-q/--quiet`|optional|Silence all logging messages.|

### fetch_regions.py
```
fetch_regions.py [-h] -l LENGTH [--verbose] [--quiet] bedfile fastafile
```

Takes a file (_BED_) of genome-wide motif occurrences and a reference genome sequence (_FASTA_) as input and generates _in silico_ reads based on motif occurrences and desired read length.

|arguments|option|description|
|---|---|---|
|`bedfile`|required|BED file containing the motif occurrences genome-wide.|
|`fastafile`|required|FASTA file containing the reference genome sequence.|
|`-l/--length`|required|Length (in bp) of the _in silico_ reads to be generated.|
|`-v/--verbose`|optional|Set level of logging messages.|
|`-q/--quiet`|optional|Silence all logging messages.|

### find_motif_occurrences.py
```
find_motif_occurrences.py [-h] filename motif
```

Takes a (gzipped) genome reference file (_FASTA_) as input and finds all positions at which the provided motif occurs. Outputs the found positions in BED format to STDOUT. Note, does not handle revcomp well. But that doesn't matter if your motif is itself palindromic.

|arguments|option|description|
|---|---|---|
| `filename`|required|Genome reference file name (FASTA).|
|`motif`|required|Motif to be found (e.g. "GATC").|

### generate_celseq_counts.py
```
generate_celseq_counts.py [-h] [--verbose] [--quiet] --gtf GTF [--mode {intersection-strict,intersection-nonempty,union}] [--min-mapq MIN_MAPQ] [--outfile HDF5_FILE] bamfile [bamfile ...]
```

Takes an alignment file (_BAM_) as input and generates a table of UMI-unique transcript counts in _HDF5_ format. Whether an alignment is considered to overlap a feature is determined by the choice of overlap mode. The available modes are modeled after the [HTseq modes](https://htseq.readthedocs.io/en/release_0.11.1/count.html).

|arguments|option|description|
|---|---|---|
|`bamfile`|required|Input BAM file(s) of aligned CEL-Seq reads.|
|`-g/--gtf`|required|Reference GTF file.|
|`-m/--mode`|required|Overlap mode for aligning reads to features. Based on HTseq. Options: {intersection-strict, intersection-nonempty, union}.|
|`--min-mapq`|optional|Minimum mapping quality of aligned reads. Reads below the threshold are discarded [default: 0].|
|`-o/--outfile`|optional|Output file (HDF5) to store UMI-unique transcript counts. If not provided, output is send to STDOUT as a table of geneid and associated counts.|
|`-v/--verbose`|optional|Set level of logging messages.|
|`-q/--quiet`|optional|Silence all logging messages.|

### generate_damid_counts.py
```
generate_damid_counts.py [-h] [--verbose] [--quiet] --outfile OUTFILE [--invalid-outfile INVALID_OUTFILE] [--min-mapq MIN_MAPQ] [--umi-length UMI_LENGTH] [--keep-n KEEP_N] [--min-editdistance MIN_EDITDISTANCE] --pos-file HDF5_FILE infile
```

Takes an alignment file (_BAM_) as input and generates a table of UMI-unique GATC counts (_HDF5_).

|arguments|option|description|
|---|---|---|
|`infile`|required|BAM input file of aligned DamID reads sorted by genomic position. Use "-" when reads are piped from STDIN.|
|`--outfile`|required|Output file (HDF5) to which UMI-unique counts are written.|
|`--invalid-outfile`|optional|Output file (BAM) to store invalid reads (i.e. too low mapping quality or not aligning to a GATC).|
|`--min-mapq`|optional|Minimum mapping quality of aligned reads. Reads below the threshold are discarded [default: 0].|
|`--keep-n`|optional|Maximum number of unique UMIs to allow per position [default: keep all].|
|`--min-editdistance`|optional|Minimum edit distance between pairs of UMIs that are observed per position [default: 1].|

### write_posarray.py
```
write_posarray.py [-h] [-l ELEMENT_LENGTH] --outfile OUTFILE pos_file
```

Takes a _BED_ file describing all motif occurrences and generates a table in HDF5 format with all motif positions genome-wide. For each chromosome, the first entry is the position of the first motif occurrence on the plus strand, subsequent values represent the distance to the next motif. Only positions on the plus strand are indicated, motif positions on the minus strand can be inferred.

|arguments|option|description|
|---|---|---|
|`pos_file`|required|BED input file with motif positions.|
|`--outfile`|required|Output file name (HDF5) for posarray.|
|`-l/--element-length`|optional|Length of the motif in bp (e.g. 4 for the GATC motif) [default: infer length from input].|

## License
**scDamAndTools** is licensed under **MIT**. The detailed license terms are in the **LICENSE** file.

## References

Original research paper describing scDam&T-seq:  
- Rooijers, K., Markodimitraki, C. M., Rang, F. J., de Vries, S. S., Chialastri, A., de Luca, K. L., Mooijman, D., Dey, S. S. & Kind, J. (2019). Simultaneous quantification of proteinâ€“DNA contacts and transcriptomes in single cells. _Nature biotechnology_, 37(6), 766.

Elaborate description of protocol and analysis:  
- Markodimitraki, C. M., Rang, F.J., ..., Kind, J. (2019). scDam&T-seq for simultaneous quantification of protein-DNA interactions and transcriptomes in single cells. (2019). _unpublished_.
